### PR TITLE
semantic-release

### PR DESCRIPTION
**Pull request by daks**
My own work on implementing semantic-release and update Kitchen tests. This is a WIP which needs to be reviewed, cleaned before any merged.
It also needs to be compared to https://github.com/saltstack-formulas/packages-formula/pull/54 which also has started work on this implementation.

Complete tests results is here https://travis-ci.com/daks/packages-formula/builds/126515555. Only OS not working are Opensuse15 _(as always? ;)_, Ubuntu 16.04, Debian 8 (old), Centos 6 (old).

CC @myii @javierbertoli

### CODE REVIEW

**Comment by myii** [comment]
@daks Just some initial feedback.

How do you feel about having commits added to this PR?  I've run this through via. `ssf-formula` and got a few changes to make, which would be easier just to add than explain.  Also made a few other modifications and `opensuse` seems to be working now:

* https://travis-ci.org/myii/packages-formula/builds/585591063

By the way, the issue with `opensuse` is that `opensuse-leap-15` shows up as `unknown` -- the way to get around it is by using `platform[:family]` + `suse` (`os[:family]` probably works fine, too -- but I recall somewhere that `platform` is the new standard).

It seems I've broken `amazon` for now but that should be a simple fix.  You can see the commit from there, it's only the initial step.

The `yamllint` errors would also need to be fixed:

```bash
packages-formula$ yamllint -s .
./.travis.yml
  21:10     error    empty value in block mapping  (empty-values)
  22:5      warning  comment not indented like content  (comments-indentation)

./pillar.example
  1:1       warning  missing document start "---"  (document-start)
  4:1       warning  comment not indented like content  (comments-indentation)
  26:1      warning  comment not indented like content  (comments-indentation)
  64:1      warning  comment not indented like content  (comments-indentation)
  65:89     error    line too long (121 > 88 characters)  (line-length)
  69:87     error    trailing spaces  (trailing-spaces)
  70:89     error    line too long (96 > 88 characters)  (line-length)
  94:43     warning  missing starting space in comment  (comments)
  97:89     error    line too long (93 > 88 characters)  (line-length)
  98:12     warning  missing starting space in comment  (comments)
  98:89     error    line too long (100 > 88 characters)  (line-length)
  101:35    warning  missing starting space in comment  (comments)
  106:89    error    line too long (102 > 88 characters)  (line-length)
  112:89    error    line too long (112 > 88 characters)  (line-length)
  118:89    error    line too long (108 > 88 characters)  (line-length)
  128:12    warning  truthy value should be one of [false, true]  (truthy)

./packages/defaults.yaml
  3:1       warning  missing document start "---"  (document-start)
  33:32     error    too few spaces after comma  (commas)
  35:14     warning  truthy value should be one of [false, true]  (truthy)
  36:23     error    too few spaces after comma  (commas)
  47:20     warning  missing starting space in comment  (comments)
  50:24     warning  missing starting space in comment  (comments)
  54:12     warning  truthy value should be one of [false, true]  (truthy)

./packages/osfamilymap.yaml
  3:1       warning  missing document start "---"  (document-start)

./packages/osmap.yaml
  1:1       warning  missing document start "---"  (document-start)
  14:23     error    too few spaces after comma  (commas)
  15:14     warning  truthy value should be one of [false, true]  (truthy)
  19:13     error    empty value in block mapping  (empty-values)
```

File: test/integration/default/controls/remote_pkgs_spec.rb [hunk 1]
```diff
@@ -1,4 +1,11 @@
 ### REMOTE PKGS
+remote_pkgs = ['zoom']
```
Comment (n-rodriguez) [inline test/integration/default/controls/remote_pkgs_spec.rb]: Not very Rubyish ^^

```ruby
remote_pkgs = 
  case os[:name]
  when 'opensuse'
    []
  else
    ['zoom']
  end
```

Comment (daks) [reply test/integration/default/controls/remote_pkgs_spec.rb]: okay for this. I just don't like `%w()` :)

**Comment by daks** [comment]
@myii no problem to you adding commits.

**Comment by daks** [comment]
just remembered this WIP PR, what is its status? is there still work to be done?

**Comment by myii** [comment]
@daks Let me run `ssf-formula` against it and add a commit here.  I'll do that over the next hour or so.  Then we can see where we're at.

**Comment by myii** [comment]
@daks Right, I've just seen that I did start this process a couple of weeks back but ran into a trouble:

* https://travis-ci.org/myii/packages-formula/builds/593265821

All of these were failing with:

```ruby
Failed to complete #verify action: [progressbar is not part of the bundle. Add it to your Gemfile.] ...
```

**Comment by myii** [comment]
Adding `progressbar` and `minitest` to the `Gemfile` seems to make a difference...

**Comment by myii** [comment]
@daks OK, for whatever reason, I had to add the "unwanted" gems as well, to get through to the `kitchen verify` stage.  But then it has issues with these gems.  I'll point you to my developments and you can continue from there:

* https://travis-ci.org/myii/packages-formula/builds/599735750 (see the individual instances for the problem):

```ruby
  ×  Wanted/Required gems: should be installed (2 failed)
     ×  true should be installed
     undefined method `installed?' for true:TrueClass
     Did you mean?  instance_of?
     ×  true should be installed
     undefined method `installed?' for true:TrueClass
     Did you mean?  instance_of?
  ×  Unwanted gems: should be uninstalled (3 failed)
     ×  true should not be installed
     undefined method `installed?' for true:TrueClass
     Did you mean?  instance_of?
     ×  true should not be installed
     undefined method `installed?' for true:TrueClass
     Did you mean?  instance_of?
     ×  true should not be installed
     undefined method `installed?' for true:TrueClass
     Did you mean?  instance_of?
```

You can access my branch to see the commits I added to get to this stage.

**Comment by myii** [comment]
@daks Had a little conversation with @javierbertoli recently and the idea is to get this PR merged in and then Javier can look at bringing in the juicy bits from #54.  Now, this has been a bit painful but I've got everything working:

* https://travis-ci.org/myii/packages-formula/builds/622280219

The diff that would be applied on top of this PR is here:

* https://github.com/myii/packages-formula/compare/943582fd99d7...de3e37d9733a
  - Of course, the commits will be cleaned up
  - Please confirm these changes are OK to add to this PR

Now for the bad news: this doesn't apply cleanly on top of the `master` branch.  There have been a number of changes since this PR first started, so there'll have to be a rebase.  I've got no problem doing that as well, as long as the changes I'm proposing to add here are OK.  After having spent the time, I'd strongly suggest to get this merged ASAP so that it doesn't all have to be done again.  With the nature of this formula, there are many things that keep moving, so it's best to get the tests in now, so that they can be updated along the way.

**Comment by daks** [comment]
@myii I'm sorry but I have not enough time these days to work on it again. 

Your proposed changes seems quite a lot, some to fix tests and also (it seems) that some are only needed to make the PR in-line with last version of `template-formula`. Maybe those last are not useful to merge this PR and can be done in a separated one.
But I'm OK with whatever needs to be done, adding commits, rebasing, squashing...

**Comment by myii** [comment]
@daks There's no other way, unfortunately.  The pre-salted images have also changed since this PR was first proposed and that affects the test, etc.  And the formula itself has moved on significantly, so that still needs to be catered for.  I'll take you on your last statement: "I'm OK with whatever needs to be done" -- because that's the stage we're at so that this PR can be recovered.  Otherwise, I'm fine with opening _another_ PR and pulling in your commits to that development.  That will probably be cleaner than working here but a third `semantic-release` PR seems a bit too much.  However, if your prefer that, then I'll do that.  I may be forced to do that if I can't rebase this PR...

**Comment by daks** [comment]
@myii no no I'm ok, go on! and thanks a lot :)

**Comment by myii** [comment]
@daks Thanks, I'll let you know how I get along... it's not like you're going to miss the major roadworks!

**Comment by myii** [comment]
@daks So here's my proposal, rebased on the latest commit in this formula:

* https://github.com/myii/packages-formula/compare/17b433126b6e%5E...f0d18b2e5a81
* https://travis-ci.org/myii/packages-formula/builds/637631361

I had to tweak some of your commits when fixing conflicts during the rebasing process.

It isn't perfect but it will get `semantic-release` in place and we can refine things from there.  If you're OK with it, I'll try to rebase this PR with those commits; not sure I'll have the permissions to do that.  It would be even better if you can grab those commits from [my branch](https://github.com/myii/packages-formula/tree/61_final) and push them here yourself.  Alternatively, the simplest solution of all is for me to open a new PR.

@noelmcloughlin I had to retouch some of your `golang` integrations to get all the tests working but those can be reintroduced afterwards if you like.

@javierbertoli I didn't manage to integrate your changes in #54, as you already know.  Hopefully, you can find a way to bring them in once `semantic-release` is up and running.

**Comment by myii** [comment]
OK, I've pushed the latest changes so this is ready to be reviewed and merged -- it only took 5 months!  Since this is work by @daks and myself, we'd appreciate someone else to have a look if possible.  Worse case, one of us will just merge it anyway!

CC: @javierbertoli @aboe76 @noelmcloughlin.

**Comment by myii** [comment]
@aboe76 Yes, that would be helpful.

**Comment by daks** [comment]
sorry for late answer @aboe76: ok for me

**Pull request merged**